### ISSUE TITLE
Minsky boots genesis kernel when mutliple interfaces have dhcp dynamic ranges.

### ISSUE THREAD

**Issue by markprez** [xcat2/xcat-core#1881]
xCAT's default response to a dhcp bootp request is to boot the genesis kernel.
If you have multiple interfaces service dhcp dynamic responses but only one setup to provision and run distributed commands to the nodes the default on the other network will be genesis boot.
To dedicate the shared BMC port to BMC traffic only, we need to prevent dhcp responses to the OS side of the shared connection (enP_f0).
If xCAT is setup to install on the enP_f1 OS network:
Object name: c712garrisonBMC
    gateway=10.128.130.1
   dynamicrange=10.128.130.100-10.128.130.140
    mask=255.255.255.0
    mgtifname=enP4p1s0f0
    net=10.128.130.0
    tftpserver=10.128.130.1
Object name: c712garrisonOS
    dynamicrange=192.168.130.100-192.168.130.140
    mask=255.255.255.0
    mgtifname=enP4p1s0f1
    net=192.168.130.0
    tftpserver=192.168.130.1
*
tabdump site | grep dhcpinterfaces
"dhcpinterfaces","enP4p1s0f0,enP4p1s0f1",,
*
BMC requests will be handled across the BMC network, but so will enP*f0 requests.
This is a requirement to prevent xCAT from trying to provision the node on the shared network.
Er Tao recommended removing /tftpboot/pxelinux.cfg/p/10.130.130.0_24 file before provisioning the nodes, this will prevent xCAT from sending the genesis kernel.

**Comment by cxhong** [comment]
Ralph suggested we should modify the dhcpd.conf too .

**Comment by daniceexi** [comment]
We may enhance the ``dhcpinterfaces`` to specify which nic cannot ``boot`` by adding a pending ``noboot`` keyword like following. If this is configured, we won't add the genesis load entry in the dhcpd.conf for this nic.

~~~~
"dhcpinterfaces","enP4p1s0f0!noboot,enP4p1s0f1",,
~~~~

**Comment by ralphbellofatto** [comment]
That looks like a good way to express this.....

What is the time line for getting this implemented...

**Comment by ralphbellofatto** [comment]
I have been talking to our engineers who are successfully setting up garisions with dual cables, I think xcat already handles this case.

The trick is to place each of the cables on connected to the garison's on a separate vlan, such that untagged traffic.  This way the untagged traffic gets directed to different ip addresses on the managemt node depending on what vlan was assigned to the switch.

The current deployed do not do a separate vlan tag for the bmc, but continue to let it function on its default vlan and asign it at the switch level...

**Comment by daniceexi** [comment]
@ralphbellofatto I think for the case that make bmc use the port 0 dedicatedly, you don't need to use tagged vlan anymore. Just two ports, each one do it's job separately.

**Comment by ralphbellofatto** [comment]
For the genesis to work correctly, eth0 and eth1 need reach the management on different NICS.  That can be done by assigning each on its own vlan, or use completely separate switches.
  Then the genesis sorts out the mac address, it and writes out the data as follows:

```
mac=70:e2:84:14:12:d7!*NOIP*|70:e2:84:14:12:d8
```

This results in the leases file containing this after discovery:

```
host ray10 {
dynamic;
hardware ethernet 70:e2:84:14:13:f5;
fixed-address 192.168.64.10;
supersede server.ddns-hostname = "ray10";
supersede host-name = "ray10";
supersede conf-file = "http://192.168.64.1/tftpboot/petitboot/ray10";
}
 

host ray10-noip70e2841413f4 {
dynamic;
hardware ethernet 70:e2:84:14:13:f4;
supersede server.allow-booting = 00;
}
```

This will result in a dynamic address being assigned to eth0 and a fixed address assigned by dhcp to eth1.

It is optional if you move the BMC to a separate vlan, as eth0 won't really be in use here, although it will receive a dynamic address from the dhcp server.  Although, since it is possible for the compute nodes to transmit data on eth0 unless it is specifically downed by a post script, having the genesis code isolate BMC network to a separate vlan is desirable.

**Comment by ralphbellofatto** [comment]
I further analyzed the system from which the above dhcp leases file and came from and i'm seeing the mac settings in the lsdef file are inconsistent, even though the person who set the system up reported them working reliably..

here are a subset of the lsdef sections that we had pulled from this system:

```
Object name: ray10
    ip=192.168.64.10
    mac=70:e2:84:14:13:f5
Object name: ray11
    ip=192.168.64.11
    mac=70:e2:84:14:11:c1
Object name: ray12
    ip=192.168.64.12
    mac=70:e2:84:14:11:8a!*NOIP*|70:e2:84:14:11:8b
Object name: ray13
    ip=192.168.64.13
    mac=70:e2:84:14:13:40!*NOIP*|70:e2:84:14:13:41
```

notice that some of the nodes identified themselves with the NOIP and some did not.  It will take some more tracing of either this system or an experimental system to figure out why this is the case.

Attached is the lsdef file from that system.
[xcatinfo.txt](https://github.com/xcat2/xcat-core/files/492101/xcatinfo.txt)

We need a solution that will allow the this system to work reliably, so getting back to the original proposal, i think we need a way for the network table to be setup such that it does NOT provide genesis images on a specific subnet.

To make this work one still needs to put each garrison cable on a separate vlan, because the first part of the dhcp discover process is a broadcast from the ethernet port, so it has to guarantee that that broadcast reaches the correct management node device.

**Comment by zet809** [comment]
We are still discussing the solution for this issue, so move to next release.

**Comment by ralphbellofatto** [comment]
We are deploying a 2 wire solution really soon now, so we need a workaround that we can at least patch into the code we have, rather than relying on our admins to remember to hand edit a file after any operation that creates genesis files or rewrites the dhcp leases file.

That is way too error prone..

A simple patch file with a temporary solution could help us.

**Comment by daniceexi** [comment]
The patch should not be hard so it can make a patch soon after we made the decision. I try to move this issue back to 2.12.3 to see whether we can make the fix in 2.12.3 release.

**Comment by ralphbellofatto** [comment]
that would be very good for us, having to remember to edit a file after certain operations makes things rather difficult.

BTW. any explanation why we saw the inconsistent discovery of mac addresses noted in the comment above?

the earlier dhcp lease telemetry seems to imply that ray10 got it right at least once and then somehow got it wrong again...

**Comment by whowutwut** [comment]
Requires documentation changes to include the fix provided by @zet809  before this issue can be put to pending, Will see if I can do that today/soon...

**Comment by whowutwut** [comment]
@zet809  The discussion I brought up last night is regarding the few comments above from @ralphbellofatto , the Issues about the mac address line in the node definition losing the NOIP one. Althought the dhcp.leases file does show at some point we are aware of the NOIP, but later on it's lost.  It could be that some admin had changed it, but trying to see if there is a scenario where genesis could be run again on the single interface and create modifications to the node definition, unaware of the 2nd MAC?  If you have some more comments on this, maybe you can add your thoughts

**Comment by zet809** [comment]
@whowutwut If the MAC with 'NOIP' doesn't appear in the mac attribute for a pre-defined node after node discovery, it means the nic with the MAC  doesn't get a dynamic IP address when generating discovery request package. The dynamic range is removed for the BMC/eth0 network?

**Comment by ralphbellofatto** [comment]
We have seen evidence that the code that handles this miss-behaves.

Inside the lsdef files for a garison system setup with 2 wire boot we see this:
```
Object name: ray10
    ip=192.168.64.10
    mac=70:e2:84:14:13:f5
Object name: ray11
    ip=192.168.64.11
    mac=70:e2:84:14:11:c1
Object name: ray12
    ip=192.168.64.12
    mac=70:e2:84:14:11:8a!*NOIP*|70:e2:84:14:11:8b
Object name: ray13
    ip=192.168.64.13
    mac=70:e2:84:14:13:40!*NOIP*|70:e2:84:14:13:41
```

However, inside the dhcp leases file we see:

```
host ray10 {
dynamic;
hardware ethernet 70:e2:84:14:13:f5;
fixed-address 192.168.64.10;
supersede server.ddns-hostname = "ray10";
supersede host-name = "ray10";
supersede conf-file = "http://192.168.64.1/tftpboot/petitboot/ray10";
}
 

host ray10-noip70e2841413f4 {
dynamic;
hardware ethernet 70:e2:84:14:13:f4;
supersede server.allow-booting = 00;
}
```

This implies that ray10 at one point HAD the noip clause and then somehow was overwritten.

Seeing this in the system has us a bit concerned.  Also, if i do read that statement in the dhcp leases file, that says the IP DOES get an ip address when the interface is activated,  It just does not get a bootp file from the dhcp server.

Since the changes to the xcat configuration are not tracked on this system, we really don't know what was done to it, only that this was left behind.

**Comment by zet809** [comment]
In fact, for a new server, there won't be a 'noip' entry in the release file. All the activated NICs may get a dynamic IP address if there is a dhcp server and dynamic range for it.
For the mac attribute that doesn't include the entry that <mac!*NOIP*>, the only reason is when sending hardware discovery request, no dynamic IP got for that NIC.
So for your issue, maybe there is not so many dynamic IP available for the BMC network, so some of the shared host NICS doesn't get a dynamic IP and then doesn't appear in the mac attribute of the matching node.

**Comment by ralphbellofatto** [comment]
If i understand the code correctly, NOP does not really mean no IP but it means this goes should go in the leases file:

supersede server.allow-booting = 00

What is being noted in this instance, is that we have indirect evidence that the NOIP setting (and hence what causes the allow-booting flag to be set) somehow got unset on some but not all of the nodes.

This system is cabled with both ports all the time, so the question here, is how did the system get into this inconsistent state..

**Comment by zet809** [comment]
@ralphbellofatto I think the possible reason is, there is no enough dynamic IP address for the shared BMC/eth0 port. Then, when doing hardware discovery, the eth0 of the host which can not get an dynamic IP, won't have "*NOIP*" appended to the mac of it for the matched xcat node definition.

**Comment by ralphbellofatto** [comment]
there are 60 nodes in this cluster. and this is the and these are the network configurations:

I'm seeing a 128 node dynamic range here, which should "barely" be enough of a dynamic range for both the BMC's and the regular nodes.

However,  once genesis is run, then the BMC's are no longer dynamic, but their addresses are burned in.

```
[root@ray1 ~]# lsdef -t network -l
Object name: BMC_NETWORK
    dhcpserver=192.168.0.128
    dynamicrange=192.168.0.129-192.168.0.254
    gateway=<xcatmaster>
    mask=255.255.255.0
    mgtifname=enP1p12s0f1
    net=192.168.0.0
    tftpserver=192.168.0.128
Object name: IBM_SITE_NETWORK
    gateway=9.10.137.250
    mask=255.255.255.0
    mgtifname=enP1p12s0f3
    net=9.10.137.0
Object name: IPoIB0_NETWORK
    gateway=<xcatmaster>
    mask=255.255.255.0
    mgtifname=ib0
    net=192.168.36.0
Object name: IPoIB1_NETWORK
    gateway=<xcatmaster>
    mask=255.255.255.0
    mgtifname=ib1
    net=192.168.37.0
Object name: PROV_NETWORK
    dynamicrange=192.168.64.129-192.168.64.254
    gateway=<xcatmaster>
    mask=255.255.255.0
    mgtifname=enP1p12s0f0
    net=192.168.64.0
    tftpserver=192.168.64.1
```

**Comment by zet809** [comment]
@ralphbellofatto Hi, ralph, we have discovery request stored in discoverydata table, will you pls send me its content to take a look at? You can use "tabdump discoverydata". Thx!

**Comment by ralphbellofatto** [comment]
here is the file from that system:

NOTE: we have not duplicated the problem yet, as we are still waiting for our research cluster to be reconfigured to use the 2 wire configuration.

This data is from another test system we have no control over, but have mined it for reference data on how a 2 wire configuration works.

So we are only noticing the state of the lsdef files noting they are inconsistent, and wondering how they got that way.

[discoverdata.zip](https://github.com/xcat2/xcat-core/files/563991/discoverdata.zip)

**Comment by zet809** [comment]
@ralphbellofatto Based on the data attached, it seems the node 'ray2,ray3,ray4,ray5,ray6,ray7' will only have 1 MAC address wrote in the mac attribute. No MAC address with'*NOIP*' appended will be wrote in mac attribute for those nodes.
The reason is for those nodes above, only 1 mac get a dynamic address during discovery process.
But from my point of view, whether there is '*NOIP*' entries or not, the booting to genesis won't be affected with our solution.

**Comment by whowutwut** [comment]
Closing as we have resolved this with the noboot option in 2.12.4 release.   Please open new issues if the problem re-appears.
